packages/providers/README.md

```diff
@@ -1,3 +1,183 @@
+import { QueryClient } from '@tanstack/react-query';
+import { PloneProvider } from '@plone/providers';
+import PloneClient from '@plone/client';
+import { ReactQueryDevtools } from '@tanstack/react-query-devtools';
+import { flattenToAppURL } from './utils';
+import config from './config';
+
+// Custom hook to unify the location object between NextJS and Plone
+function useLocation() {
+  let pathname = usePathname();
+  let search = useSearchParams();
+
+  return {
+    pathname,
+    search,
+    searchStr: '',
+    hash: (typeof window !== 'undefined' && window.location.hash) || '',
+    href: (typeof window !== 'undefined' && window.location.href) || '',
+  };
+}
+
+const Providers: React.FC<{
+  children?: React.ReactNode;
+}> = ({ children }) => {
+  // Creating the clients at the file root level makes the cache shared
+  // between all requests and means _all_ data gets passed to _all_ users.
+  // Besides being bad for performance, this also leaks any sensitive data.
+  // We use this pattern to ensure that every client gets its own clients
+  const [queryClient] = React.useState(
+    () =>
+      new QueryClient({
+        defaultOptions: {
+          queries: {
+            // With SSR, we usually want to set some default staleTime
+            // above 0 to avoid refetching immediately on the client
+            staleTime: 60 * 1000,
+          },
+        },
+      }),
+  );
+
+  const [ploneClient] = React.useState(() =>
+    PloneClient.initialize({
+      apiPath: config.settings.apiPath,
+    }),
+  );
+
+  const router = useRouter();
+
+  return (
+    <PloneProvider
+      ploneClient={ploneClient}
+      queryClient={queryClient}
+      // NextJS doesn't have a useLocation hook, so we need to unify this
+      // in a custom hook
+      useLocation={useLocation}
+      navigate={(to) => {
+        router.push(to);
+      }}
+      useParams={useParams}
+      useHref={(to) => flattenToAppURL(to)}
+      flattenToAppURL={flattenToAppURL}
+    >
+      {children}
+      <ReactQueryDevtools initialIsOpen={false} />
+    </PloneProvider>
+  );
+};
+
+export default Providers;
+
+```
+
+You can use anywhere in your app by using the hook `usePloneProvider`.
+
+```tsx
+import { usePloneProvider } from '@plone/providers';
+
+const { ploneClient } = usePloneProvider()
+```
+
+or any other context property.
+
+```tsx
+const { navigate } = usePloneProvider()
+```
+
+### `PloneClientProvider`
+
+`PloneProvider` in a group of other smaller providers.
+In case you need them separatedly, you can also instantiate and use them standalone.
+However, only do this is the framework has some limitation on using the bulk `PloneClientProvider`.
+
+```ts
+export type PloneClientProviderProps = {
+  client: InstanceType<typeof PloneClient>;
+  queryClient: QueryClient;
+  children?: React.ReactNode;
+};
+```
+
+and the related hook:
+
+```tsx
+import { usePloneClient } from '@plone/providers';
+
+const client = usePloneClient()
+```
+
+or
```

Comment:
```suggestion
Alternatively, you can use the following hook.
```

Reply:
in this case, it's another usage use case, not that it's alternative.

Reply:
I'm not sure what to write for this section. I have a hard time translating the code to narrative text. It sounds like the pattern is the following.

````
The following snippets show two use cases.

You can instantiate the provider with its hook.

```ts
// instantiation
export type PloneClientProviderProps = {
  client: InstanceType<typeof PloneClient>;
  queryClient: QueryClient;
  children?: React.ReactNode;
};

// its related hook
import { usePloneClient } from '@plone/providers';
const client = usePloneClient()
```
````